### ISSUE TITLE
Recuperar senha

### ISSUE THREAD

**Issue by pedrombl** [servicosgovbr/editor-de-servicos#404]
Funcionalidade para usuário recuperar senha de autenticação. Fluxo básico:

1. Usuário esquece senha
2. Usuário comunica Ponto Focal 
3. Ponto focal requisita nova senha utilizando funcionalidade do Editor
4. Um link com um identificador único e com validade limitada é gerado e usuário fica desabilitado
5. Ponto focal informa usuário do link 
6. Usuário acessa link gerado anteriormente e uma nova senha é cadastrada e fica habilitado

Tarefas
- [x] Criar tabela no banco para armazenar token/usuario
- [x] Troca senha após utilizar token e colocar CPF
- [x] Criar tela para mostrar token e informações basicas do usuário
- [x] Criar tela para recuperar senha
- [x] Invalida token após 3 tentativas erradas
- [x] Invalida token após 24 horas

**Comment by gustavo-freitas** [comment]
Considerando que não enviaremos senhas por e-mail, pensei nos seguintes passos (utilizando a lógica da criação de usuários):

1) Usuário esquece senha
2) Usuário requisita nova senha utilizando e-mail 
3) E-mail informativo é enviado para usuário, explicando que ele/a pode acessar novamente o editor
4) No primero acesso do usuário, depois da requisição de nova senha, usuário é forçado a cadastrar nova senha

**Comment by almeidafab** [comment]
Sim @gustavo-freitas faz sentido

**Comment by gustavo-freitas** [comment]
Ok @almeidafab. Ainda estamos trabalhando como vai ser exatamente esse processo. @sahrosa @pedrombl @gustavo-freitas trabalhando nesse processo.

**Comment by gomex** [comment]
@gustavo-freitas quando você fala "No primeiro acesso do usuário, depois da requisição de nova senha, usuário é forçado a cadastrar nova senha" isso quer dizer que ele precisa clicar no link que ele recebe por e-mail pra que a obrigatoriedade passe "a funcionar", certo?

**Comment by gustavo-freitas** [comment]
@gomex sim, essa era a ideia inicial. Entretanto, não iremos trabalhar mais com e-mails automáticos. @sahrosa pode te passar um contexto melhor sobre a nova solução que pensamos. De qualquer forma, esse cartão deve ser atualizado logo com esses detalhes.

**Comment by sahrosa** [comment]
Do ponto de vista da interface para os usuários, a recuperação de senha será apenas uma mensagem, conforme imagem abaixo:
 ![screen shot 2015-11-18 at 6 23 26 pm](https://cloud.githubusercontent.com/assets/2932644/11275147/611e1050-8ec3-11e5-842a-27612c273ca5.png)

**Comment by mformiga** [comment]
@sahrosa O ideal seria enviar a mensagem de recuperação de senha para o e-mail cadastrado mas dentro do cenário de tempo, concordamos em implementar dessa forma. Favor prosseguir.

**Comment by sahrosa** [comment]
Legal @mformiga. 

@gustavo-freitas, tu quer complementar este cartão com a questão mais técnica ou posso complementar?

**Comment by gustavo-freitas** [comment]
@mformiga @almeidafab novo fluxo atualizado na descrição principal do cartão.
@sahrosa pode me ajudar a pensar nas telas de gerar nova senha (ponto focal) e recuperar senha (usuário)? Seguem ideias "iniciais":

**Gerar nova senha (lista de usuários que o ponto focal vai ver)**

![image](https://cloud.githubusercontent.com/assets/2887992/11285515/fb5ed2ac-8ef6-11e5-9c05-fc087f0f7617.png)

**Recuperar senha (usuário acessa link gerado pelo ponto focal)**

![image](https://cloud.githubusercontent.com/assets/2887992/11285622/a32f0b14-8ef7-11e5-9bd8-9314a757c673.png)

**Comment by mformiga** [comment]
@gustavo-freitas está faltando o campo para busca de usuários.

**Comment by gustavo-freitas** [comment]
@mformiga nossa combinação foi de não ter uma busca nessa funcionalidade, pois iria adicionar complexidade e, consequentemente, escopo. A @sahrosa está fazendo mockups mais detalhados e poderemos ter uma ideia melhor de como ficarão essas telas.

**Comment by sahrosa** [comment]
A funcionalidade de recuperação de senha será acessada através do ícone de chave na lista de usuários, conforme imagem abaixo:
![screen shot 2015-11-23 at 11 06 42 am](https://cloud.githubusercontent.com/assets/2932644/11337230/8a7d45a8-91d2-11e5-8874-307eee0cad9b.png)

Ao acessar a área de recuperação de senha, o admin/ponto focal poderá copiar o link com token para criação de nova senha, conforme imagem abaixo: 
![screen shot 2015-11-23 at 11 07 26 am](https://cloud.githubusercontent.com/assets/2932644/11337276/e824094e-91d2-11e5-8642-96a4ad929697.png)

O usuário que solicitou ao admin/ponto focal uma nova senha receberá um link para a seguinte página, onde uma nova senha poderá ser cadastrada:
![screen shot 2015-11-23 at 11 07 42 am](https://cloud.githubusercontent.com/assets/2932644/11337307/28843b8a-91d3-11e5-9e9a-ea4fd9cee70b.png)

**Comment by gustavo-freitas** [comment]
@sahrosa ficou ótimo =D

**Comment by almeidafab** [comment]
@gustavo-freitas @sahrosa fiz um teste agora de manhã e não consegui alterar a senha. Apareceu uma mensagem dizendo que o CPF informado não é compatível com o cadastrado. Mas eu estava colocando o cpf correto

**Comment by sahrosa** [comment]
Oi @almeidafab, você estava testando em qual ambiente? 

Estou tentando reproduzir aqui mas aparentemente tudo certo! Vamos te avisando se encontrarmos algum comportamento parecido.

**Comment by almeidafab** [comment]
@sahrosa no ambiente de desenvolvimento.
Quando acesso a tela de recuperar senha:

![captura de tela de 2015-12-04 16 02 04](https://cloud.githubusercontent.com/assets/9214263/11597089/7c411c90-9aa0-11e5-869e-eda970ec72b2.png)

Abro o link e preencho o cpf e a nova senha aparece a seguinte mensagem:

![captura de tela de 2015-12-04 16 04 26](https://cloud.githubusercontent.com/assets/9214263/11597145/dfdeda80-9aa0-11e5-8205-422514613fbd.png)

**Comment by gustavo-freitas** [comment]
@almeidafab @sahrosa tentei reproduzir o problema em ambiente de desenvolvimento criando um novo usuário de teste (nome: usuário teste) e recuperando a senha inúmeras vezes. Tens como criar um novo usuário e tentar reproduzir novamente esse cenário?

**Comment by sahrosa** [comment]
Oi @almeidafab @gustavo-freitas, vou fechar este cartão pois acredito que este problema tinha relação com algumas instabilidades que estávamos passando. Tentamos reproduzir este problema diversas formas e por agora está tudo certo!

Se necessário, reabriremos este cartão :)

**Comment by sahrosa** [comment]
Oi @almeidafab @gustavo-freitas, vamos dar este problema como resolvido? Acredito que ele tinha relação com algumas instabilidades que estávamos passando, pois tentamos reproduzi-lo de diversas formas e por agora parece estar tudo certo! 

Podemos passar este cartão para produção?
